sdks/fabric/interoperation-node-sdk/src/AssetManager.ts

```diff
@@ -0,0 +1,523 @@
+    return assetExchangeAgreement.serializeBinary().toString();
+}
+
+// Create an asset lock structure
+function createAssetLockInfoSerialized(hashValue, expiryTimeSecs)
+{
+    const lockInfoHTLC = new assetLocksPb.AssetLockHTLC();
+    lockInfoHTLC.setHashbase64(hashValue);
+    lockInfoHTLC.setExpirytimesecs(expiryTimeSecs);
+    lockInfoHTLC.setTimespec(assetLocksPb.AssetLockHTLC.TimeSpec.EPOCH)
+    const lockInfoHTLCSerialized = lockInfoHTLC.serializeBinary();
+    const lockInfo = new assetLocksPb.AssetLock();
+    lockInfo.setLockmechanism(assetLocksPb.LockMechanism.HTLC);
+    lockInfo.setLockinfo(lockInfoHTLCSerialized);
+    return lockInfo.serializeBinary().toString();
+}
+
+// Create an asset claim structure
+function createAssetClaimInfoSerialized(hashPreimage)
+{
+    const claimInfoHTLC = new assetLocksPb.AssetClaimHTLC();
+    claimInfoHTLC.setHashpreimagebase64(Buffer.from(hashPreimage).toString('base64'));
+    const claimInfoHTLCSerialized = claimInfoHTLC.serializeBinary();
+    const claimInfo = new assetLocksPb.AssetClaim();
+    claimInfo.setLockmechanism(assetLocksPb.LockMechanism.HTLC);
+    claimInfo.setClaiminfo(claimInfoHTLCSerialized);
+    return claimInfo.serializeBinary().toString();
+}
+
+// Create a SHA-256 hash over an ASCII string
+function createSHA256HashBase64(preimage)
+{
+    return crypto.createHash('sha256').update(preimage).digest('base64');
+}
+
+
+/**
+ * First/second step of a Hashed Time Lock Contract
+ * - Lock a unique asset instance using a hash
+ **/
+const createHTLC = async (
+    contract: Contract,
+    assetType: string,
+    assetID: string,
+    recipientECert: string,
+    hashValue: string,
+    expiryTimeSecs: number,
+): Promise<{ preimage: any; result: any }> => {
+
+    if (!contract)
+    {
+        logger.error("Contract handle not supplied");
+        return { preimage: "", result: false };
+    }
+    if (!assetType)
+    {
+        logger.error("Asset type not supplied");
+        return { preimage: "", result: false };
+    }
+    if (!assetID)
+    {
+        logger.error("Asset ID not supplied");
+        return { preimage: "", result: false };
+    }
+    if (!recipientECert)
+    {
+        logger.error("Recipient ECert not supplied");
+        return { preimage: "", result: false };
+    }
+
+    let defaultExpiryTimeSecs = 0, preimage = "";
+    if (hashValue && hashValue.length > 0)
+    {
+        defaultExpiryTimeSecs = 5 * 60     // 5 mins: this is the 't' used to time out the second contract of the HTLC pair
+    }
+    else
+    {
+        defaultExpiryTimeSecs = 10 * 60     // 10 mins: this is the '2t' used to time out the first contract of the HTLC pair
+
+        // Create a random preimage of 20 bytes
+        for (let i = 0 ; i < 20 ; i++)
+        {
+            preimage += String.fromCharCode(Math.floor(Math.random() * 256))
+        }
+
+        // Hash the preimage
+        hashValue = createSHA256HashBase64(preimage);
+    }
```

Comment:
Yes, we can allow the user to submit a preimage, and I can add a parameter for that. In case a user doesn't supply it, I can use a random generation function (though I know you prefer throwing an error.)

But are there any cons to the approach that is implemented in the above code? The value of the preimage is limited to its ability to validate a claim by either party, no? So does it matter who generates it and where? It seems to me providing this functionality in code reduces one more thing the user needs to do. After all, the user (or an agent in the app) is simply interested in effecting an asset exchange with integrity, and the preimage is as important here (I think) as TLS keys would be in a typical secure web connection (where the endpoints should only care about the data getting exchanged and not the mechanism by which channel confidentiality is maintained.)

Reply:
Right, I am not opposed to offering a convenient way for a user to generate a preimage. As noted above, I think that would be useful to do. However, given this is an important parameter, I think we want to make sure that we minimize our assumptions about the user's intents. Inferring that a user wants an automated preimage created if the parameter was empty feels a bit risky. Such a scenario could be a result of a trivial bug in the calling context and not an explicit intent by the user. Hence, why I'd recommend that we either make the pre-image parameter mandatory and thus force the calling context to first generate a pre-image using our generator (or a user-supplied one) before calling this API, or we offer an alternate function that makes clear that a generated pre-image will be used. These approaches can ensure that the calling context is clear about its intentions.

Reply:
The function is already returning a generated preimage to the user, so I don't think there will be any ambiguity in the calling program about what preimage was used to finally generate the hash. I don't think we need a separate function. If the return value is empty and there is no error, the caller knows its the supplied preimage that was used.

In fact, the Fabric CA client-server registration protocol follows this pattern. When a registrar tries to enroll a new user, it gets back a response with a user ID and an auto-generated secret (password). From a security perspective, as long as the channel is confidential, I think this passes muster. In any case, it's going to be a simple function call in our scenario.